Simplify (quite a bit...) _preprocess_data

anntzer opened the pull request:
## PR Summary


Public API change: `step` no longer defaults to using `y` as
label_namer.  This is consistent with other functions that wrap `plot`
(`plot` itself, `loglog`, etc.).  (Alternatively, we could make all
these functions use `y` as label_namer; I don't really care either way.)

The plot-specific data kwarg logic was moved to
`_process_plot_var_args`, dropping the need for
general callable `positional_parameter_names`,
`_plot_args_replacer`, and `positional_parameter_names`.
`test_positional_parameter_names_as_function` and tests using
`plot_func_varargs` were removed as a consequence.

`replace_all_args` can be replaced by making `replace_names=None`
trigger replacement of all args, even the "unknown" ones.  There was
no real use of "replace all known args but not unknown ones" (even if
there was, this can easily be handled by explicitly listing the args in
replace_names). `test_function_call_with_replace_all_args` was removed
as a consequence.

~~`replace_names` no longer complains if some argument names it is given
are not present in the "explicit" signature, as long as the function
accepts `**kwargs` -- because it may find the arguments in kwargs
instead.~~ edit: not anymore, we still error out in that case.

label_namer no longer triggers if `data` is not passed (if the
argument specified by label_namer was a string, then it is
likely a categorical and shouldn't be considered as a label
anyways). `test_label_problems_at_runtime` was renamed to
`test_label_namer_only_if_data` and modified accordingly.

Calling data-replaced functions used to trigger RuntimeError in some
cases of mismatched arguments; they now trigger TypeError similarly to
how normal functions do (`test_more_args_than_pos_parameters`).

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented:
> (Alternatively, we could make all
> these functions use y as label_namer; I don't really care either way.)

I would rather do this, it seems like an oversight that we did not do this for plot and friends originally.

tacaswell commented:
This is easier to review with https://github.com/matplotlib/matplotlib/pull/10928/files?w=1  Most of the whitespace change is due to using partial instead of an extra layer of inner function to create the paramterized decorator.

tacaswell commented:
> Calling data-replaced functions used to trigger RuntimeError in some
> cases of mismatched arguments; they now trigger TypeError similarly to
> how normal functions do (test_more_args_than_pos_parameters).

Please document this as an API change.

It seems a good fraction of the simplification comes from using `signature.bind` and not re-implementing functional that core python provides?

The docstring and signature for `plot` and friends needs to be updated as they are no longer done by the decorator.

Over all :+1: !

anntzer commented:
The simplications come from
1. don't do the work that signature.bind does
2. move the plot-specific logic to its own implementation rather than attempt to have a single overgeneralized version
3. dropping replace_all_args and simplifying the replace_names logic (either replace this, this, this and this named argument, or replace all of them) to restrict them to what we need them to do

anntzer commented:
Question: Currently,
```
plot([1, 2], "foo", data={"foo": "y"})
```
works as expected but
```
plot([1, 2], [3, 4], "foo", data={"foo": "y"})
```
crashes ("unrecognized character f in format string").  This PR makes both forms "work", although I'm tempted to make *neither* work instead (I think the shorthand format string should, well, be a shorthand format string).  Thoughts?

anntzer commented:
The data kwarg actually already has its own documentation in the docstring of `plot`:
```
        data : indexable object, optional
            An object with labelled data. If given, provide the label names to
            plot in *x* and *y*.

            .. note::
                Technically there's a slight ambiguity in calls where the
                second label is a valid *fmt*. `plot('n', 'o', data=obj)`
                could be `plt(x, y)` or `plt(y, fmt)`. In such cases,
                the former interpretation is chosen, but a warning is issued.
                You may suppress the warning by adding an empty format string
                `plot('n', 'o', '', data=obj)`.
```
The same note, without the PS, is present in step; it is not present at all in loglog and friends.  While these could be rationalized I think this is out of the scope of this PR.

Added label_namer support for plot and friends.  While doing so, moved from using the terribly named `get_label` to a private `_label_from_arg` which checks whether its second argument is a string, and, if not so, defaults to None instead (to share that logic between the general and plot-specific implementations).

timhoffm commented:
Any chances that this will go in soon? Otherwise I propose to preliminary merge #10872 to fix #11389.

anntzer commented:
rebased

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -4533,7 +4533,7 @@ def barbs(self, *args, **kw):
     # Uses a custom implementation of data-kwarg handling in
     # _process_plot_var_args.
-    def fill(self, *args, **kwargs):
+    def fill(self, *args, data=None, **kwargs):
```
`data` should be added to the Parameters section in the docstring

anntzer replied:
done

jklymak commented:
What is the status of this and #10872.  Should #10872 go in as a stopgap?

anntzer commented:
I consider this ready to go in.

jklymak commented on lib/matplotlib/__init__.py:
```diff
@@ -1541,43 +1550,33 @@ def _replacer(data, key):
-def _preprocess_data(replace_names=None, replace_all_args=False,
-                     label_namer=None, positional_parameter_names=None):
+def _preprocess_data(func=None, *, replace_names=None, label_namer=None):
     """
```
I think this could benefit from a more complete docstring.  I know its internal, but I don't get a complete picture of what the preprocessing does anywhere, so its hard to understand whats going on or what this decorator/function will do.

anntzer replied:
done

jklymak commented on lib/matplotlib/__init__.py:
```diff
@@ -1588,214 +1587,107 @@ def foo(ax, *args, **kwargs)
     .. note:: decorator also converts MappingView input data to list.
     """
+
+    if func is None:
```
This could use an explanatory comment...

anntzer replied:
Done, see also discussion starting at https://github.com/matplotlib/matplotlib/pull/11398#discussion_r197648946.

jklymak commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -4605,6 +4552,10 @@ def fill(self, *args, **kwargs):
 
+        data : indexable object, optional
+            An object with labelled data. If given, provide the label names to
+            plot in *x* and *y*.
```
This could use an 'e.g.' `ax.fill(x='time', y='signal', data=thedata)`

anntzer replied:
done

jklymak commented on lib/matplotlib/axes/_base.py:
```diff
@@ -167,23 +165,67 @@ def set_prop_cycle(self, *args, **kwargs):
-
             if self.axes.name == 'polar':
                 xunits = kwargs.pop('thetaunits', xunits)
+            if xunits != self.axes.xaxis.units:
```
?? whats this for?  Seems orthogonal to the PR, and I don't think we've agreed on the units handling.  Not saying its wrong, but...

anntzer replied:
doesn't really matter (was just shuffling a few lines around to make things more compact) and reverted it

anntzer commented:
all comments handled

timhoffm commented on lib/matplotlib/__init__.py:
```diff
@@ -1525,261 +1534,152 @@ def _replacer(data, key):
     docstring : str
         The input docstring.
-    replace_names : list of strings or None
+    replace_names : List[str] or None
```
According to numpydoc this shoud be `list of str`. I don't want to start a big discussion about this because it's a private function anyway, but we've been using `list of ...` everywhere so far.

anntzer replied:
I think standard type annotations actually work quite well in cases like this but sure, let's decide this in some separate discussion.

timhoffm commented on lib/matplotlib/axes/_base.py:
```diff
@@ -180,8 +178,56 @@ def __call__(self, *args, **kwargs):
+
+        data = kwargs.pop("data", None)
+        if data is not None:
+            replaced = [mpl._replacer(data, arg) for arg in args]
```
This block is quite long. Can it be extracted into a private method?

anntzer replied:
It's really not that long (half of it is error messages and comments explaning what's happening).  Also it's modifying args and kwargs in-place, which, sure, is also possible from a helper method, but I'm not convinced it helps with legibility.  (Literally, `__call__` does three things: 1) apply units, 2) handle the "data" kwarg, 3) grab two-or-three args at a time and pass them to the plotting functoin.  We may as well keep them in that order.)

timhoffm replied:
Ok, I agree that this can stay a single method. However, it definitively needs this explanation:

> Literally, __call__ does three things: 1) apply units, 2) handle the "data" kwarg, 3) grab two-or-three args at a time and pass them to the plotting functoin.

If that was in the code, it would already be much more understandable. Can you please add that to the docstring, or alternatively split it into comments before each of the sections.

anntzer replied:
done

timhoffm commented on lib/matplotlib/axes/_base.py:
```diff
@@ -180,8 +178,56 @@ def __call__(self, *args, **kwargs):
+                    replaced[label_namer_idx], args[label_namer_idx])
+            args = replaced
+
+        while args:
```
This loop could use a short explanation.

anntzer replied:
This was literally copied from _grab_next_args in the earlier version.

anntzer replied:
done

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -1496,19 +1496,28 @@ def test(verbosity=None, coverage=False, switch_backend_warn=True,
 
-def _replacer(data, key):
+def _replacer(data, value):
     """Either returns data[key] or passes data back. Also
```
sure

anntzer commented:
Added a commit that simplifies and better documents label_namer.

anntzer commented on lib/matplotlib/tests/test_preprocess_data.py:
```diff
@@ -75,36 +58,6 @@ def func_no_ax_args(*args, **kwargs): pass
     with pytest.raises(AssertionError):
         _preprocess_data(label_namer="z")(func_args)
 
-    # but "ok-ish", if func has kwargs -> will show up at runtime :-(
```
Deleted these as we don't support "implicit" label_name ("hoping that they'll show up in kwargs") anymore (this was basically present only to support `plot()` and `fill()` which don't have `y` explicitly in the signature but always have a `y` argument; this PR changes these functions to have their own parser).

anntzer commented on lib/matplotlib/tests/test_preprocess_data.py:
```diff
@@ -205,42 +158,6 @@ def func_replace_all(ax, x, y, ls="x", label=None, w="NOT"):
         "x: [1, 2], y: [8, 9], ls: x, w: xyz, label: text")
 
-    @_preprocess_data(label_namer="y")
-    def func_varags_replace_all(ax, *args, **kwargs):
```
See above re: "implicit" label_namer.  Otherwise this test function is redundant with func_replace_all just above.

anntzer commented:
As a side point, after this PR, we may want to consider the possibility of (... being very prudent here :) ...) making _preprocess_data part of the public API (perhaps under a better name, e.g. process_data_kwarg, up to bikeshedding).
This will both allow others to reuse the functionality, and possibly to move out the spectral plotting functions out to their own mpl_toolkit :p (like everyone else they depend on _preprocess data to massage their arguments).

tacaswell commented:
Sold on this in general and on making `preprocess_data` public.

anntzer commented:
rebased

timhoffm commented:
> Sold on this in general and on making preprocess_data public.

I'm a bit hesitant about this. `_preprocess_data` seems still a bit too clever to be released in the wild.
1. I'd rather strip out the signature manipulation and instead simply require that the wrapped function has a `data` parameter.
2. Also It shouldn't manipulate the docstring. We cannot assume that users would want the data note as we have it now. - Actually I'm not a fan of this for the matplotlib docs as well. The `parameter` should appear in the `Parameters` or `Additional Parameters` section like any other parameter.

anntzer commented:
> I'd rather strip out the signature manipulation and instead simply require that the wrapped function has a data parameter.

But the inner function actually *doesn't* take `data` as parameter.

> Also It shouldn't manipulate the docstring. We cannot assume that users would want the data note as we have it now. - Actually I'm not a fan of this for the matplotlib docs as well. The parameter should appear in the Parameters or Additional Parameters section like any other parameter.

This has basically been discussed to death in the other docstring manipulation issues.

In any case this PR is not making the function public; let's keep that discussion for when the PR shows up (if ever).

timhoffm commented:
> > I'd rather strip out the signature manipulation and instead simply require that the wrapped function has a data parameter.

> But the inner function actually doesn't take data as parameter.

Yep, sorry. I was just repeating a thought from long time ago. The above comment was not the full story. Essentially, when stripping off all the fancy manipulation magic of signatures and docstrings, this could be a simple function `x, y = preprocess_data(x, y, data=data)`. But as you said, let's keep that discussion out of here.

anntzer commented:
Added a commit to make plot() error out when x/y are passed as kwargs (this silently does nothing currently, xref https://github.com/matplotlib/matplotlib/issues/12106).

tacaswell commented:
Thanks @anntzer !  Sorry for the slow review on this.
